### PR TITLE
Implement most SleepingEvents.

### PR DESCRIPTION
**Pull request by JBYoshi**
I couldn't implement `SleepingEvent.Post` with a Forge event mixin because it requires more data than `PlayerWakeUpEvent` gives, but I needed to use code more complex than `@Inject` allows, and Forge heavily modifies this method. `SleepingEvent.Tick` was implemented in SpongePowered/SpongeCommon#229.

Requires SpongePowered/SpongeCommon#283.

### CODE REVIEW

**Comment by JBYoshi** [comment]
Closing and reopening to have Travis rebuild the PR - the latest build failed because of a "Connection reset".

**Comment by JBYoshi** [comment]
Status?

File: src/main/java/org/spongepowered/mod/mixin/core/entity/player/MixinEntityPlayer.java [hunk 1]
```diff
@@ -50,4 +67,48 @@ public EntityItem dropOneItem(boolean dropAll) {
     public EntityItem dropPlayerItemWithRandomChoice(ItemStack itemStackIn, boolean unused) {
         return this.dropItem(itemStackIn, false, false);
     }
+
+    @Overwrite
```
Comment (bloodmc) [inline src/main/java/org/spongepowered/mod/mixin/core/entity/player/MixinEntityPlayer.java]: This should be moved to SpongeCommon. As for the PlayerWakeupEvent, just create a mixin to fire it before line 94.

Comment (JBYoshi) [inline src/main/java/org/spongepowered/mod/mixin/core/entity/player/MixinEntityPlayer.java]: Actually, Forge changes more than just firing the event - it adds methods like `isBed()` and `setBedOccupied()`.

File: src/main/java/org/spongepowered/mod/mixin/core/entity/player/MixinEntityPlayer.java [hunk 2]
```diff
@@ -24,21 +24,38 @@
...
 
 @Mixin(EntityPlayer.class)
 public abstract class MixinEntityPlayer extends MixinEntityLivingBase {
 
     @Shadow public InventoryPlayer inventory;
+    @Shadow public BlockPos playerLocation;
+    @Shadow protected boolean sleeping;
+    @Shadow private int sleepTimer;
     @Shadow public abstract EntityItem dropItem(ItemStack droppedItem, boolean dropAround, boolean traceItem);
+    @Shadow public abstract void setSize(float width, float height);
```
Comment (bloodmc) [inline src/main/java/org/spongepowered/mod/mixin/core/entity/player/MixinEntityPlayer.java]: setSize is not part of the target class EntityPlayer but rather Entity so this needs to be removed.

**Comment by JBYoshi** [comment]
Now requires SpongePowered/SpongeCommon#283.

**Pull request closed without merging**